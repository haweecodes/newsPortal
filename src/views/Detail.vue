<template>
  <article>
    <v-btn @click="returnToHome" icon>
      <v-icon>mdi-arrow-left</v-icon>
    </v-btn>
    <h1 class="text-center ma-5">{{ getHeadlineDetail.title }}</h1>
    <p class="text-center ma-5">
      <strong>Written By:{{ getHeadlineDetail.author }}</strong>
    </p>
    <p class="text-center ma-5">
      <strong>Published At: {{ getHeadlineDetail.publishedAt | transformDate }}</strong>
    </p>

    <v-img
      :alt="getHeadlineDetail.title"
      aspect-ratio="3"
      contain
      :src="getHeadlineDetail.urlToImage"
    ></v-img>

    <p class="ma-5">{{ getHeadlineDetail.content }}</p>
  </article>
</template>

<script>
export default {
  name: "Detail",
  computed: {
    getHeadlineDetail: function () {
      return this.$store.state.headlineDetail;
    },
  },
  methods: {
    returnToHome() {
      this.$router.push("/");
    },
  },
};
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped>
</style>
