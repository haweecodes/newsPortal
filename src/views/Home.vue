<template>
  <section>
    <v-container fluid>
      <SearchAndFilter />

      <HeadlineEditModal
        v-if="showModal"
        :open="showModal"
        @onHeadlineChange="onHeadlineChange"
        @onModalClose="toogleModal"
      />

      <LoaderWrapper :loading="getLoadingState">
        <template v-slot:content>
          <v-row v-if="getHeadlineList.length > 0" row wrap>
            <v-col
              lg="3"
              md="6"
              sm="6"
              xs="12"
              v-for="(headline, index) in getHeadlineList"
              :key="headline.url"
            >
              <HeadlineBox>
                <template v-slot:HeadlineTitle>
                  <v-tooltip bottom>
                    <template v-slot:activator="{ on, attrs }">
                      <v-btn
                        @click="editHeadline(headline, index)"
                        v-bind="attrs"
                        v-on="on"
                        color="primary"
                        fab
                        small
                        absolute
                        top
                        right
                        ><v-icon>mdi-pencil</v-icon>
                      </v-btn>
                    </template>
                    <span>Edit headline</span>
                  </v-tooltip>

                  <!-- <v-btn text @click="redirectToDetailPage(headline)"> -->
                  <p @click="redirectToDetailPage(headline)" class="link">
                    {{ headline.title }}
                  </p>
                  <!-- </v-btn> -->
                </template>
                <template v-slot:HeadlineContent>
                  {{ headline.description }}
                  <span
                    @click="redirectToOriginalPage(headline)"
                    class="link read-more"
                    >Read more...</span
                  >
                </template>
              </HeadlineBox>
            </v-col>
          </v-row>
          <v-row v-else>
            <v-col>
              <p class="text-center">No data found</p>
            </v-col>
          </v-row>
        </template>
      </LoaderWrapper>

      <HistoryDrawer v-if="drawer" :drawer="drawer" />

      <v-speed-dial
        fixed
        bottom
        right
        open-on-hover
        transition="slide-y-reverse-transition"
      >
        <template v-slot:activator>
          <v-btn color="blue darken-2" dark fab>
            <v-icon> mdi-apps </v-icon>
          </v-btn>
        </template>
        <v-btn fab dark small color="indigo" @click.stop="toggleDrawer">
          <v-icon>mdi-history</v-icon>
        </v-btn>
        <v-btn fab dark small color="red" @click="showError"> 404 </v-btn>
      </v-speed-dial>

      <v-snackbar
        v-model="showAlert"
        :timeout="-1"
        :value="true"
        absolute
        centered
        right
        top
        tile
        color="red accent-2"
      >
        {{ errorMessage }}

        <template v-slot:action="{ attrs }">
          <v-btn text v-bind="attrs" @click="showAlert = false"> Close </v-btn>
        </template>
      </v-snackbar>
    </v-container>
  </section>
</template>

<script>
import HeadlineBox from '@/components/HeadlineBox.vue';
import SearchAndFilter from '@/components/SearchAndFilter.vue';
import HeadlineEditModal from '@/components/HeadlineEditModal.vue';
import LoaderWrapper from '@/components/LoaderWrapper.vue';
import HistoryDrawer from '@/components/HistoryDrawer.vue';
import { setHistoryToLocalStorage } from '@/utils/helper';

export default {
  name: 'Home',
  components: {
    HeadlineBox,
    SearchAndFilter,
    HeadlineEditModal,
    LoaderWrapper,
    HistoryDrawer,
  },
  data: () => ({
    loading: true,
    showModal: false,
    currentHeadline: '',
    editedHeadline: '',
    editableHeadlineIndex: null,
    errorMessage: '',
    showAlert: '',
    drawer: false,
    historyList: [],
  }),
  computed: {
    getLoadingState() {
      return this.$store.state.loading;
    },
    getHeadlineList() {
      const { sourceFilter } = this.$store.state;
      return this.$store.state.newsHeadlineList.filter((headline) => {
        if (sourceFilter) return headline.source.id === sourceFilter;

        return headline;
      });
    },
  },
  beforeMount() {
    if (this.$store.state.newsHeadlineList.length < 1) {
      this.$store.commit('setLoadingState', true);
      this.$store.dispatch('fetchHeadlineList');
    }
  },
  methods: {
    redirectToDetailPage(headline) {
      this.$store.commit('setViewHeadlineDetail', headline);
      this.$router.push('/detail');
      setHistoryToLocalStorage(headline);
    },
    redirectToOriginalPage(headline) {
      setHistoryToLocalStorage(headline);
      window.open(headline.url, '_blank');
    },
    editHeadline(headline, index) {
      this.editableHeadlineIndex = index;
      this.currentHeadline = headline.title;
      this.toogleModal();
    },
    async onHeadlineChange(title) {
      this.$store.commit('editViewHeadline', {
        title,
        index: this.editableHeadlineIndex,
      });
      this.toogleModal();
    },
    toogleModal() {
      this.showModal = !this.showModal;
    },

    async showError() {
      const response = await this.$store.dispatch('wrongApiCall');
      this.alertModalConfig(response);
    },

    alertModalConfig(message) {
      this.showAlert = true;
      this.errorMessage = message;
    },
    toggleDrawer() {
      this.drawer = !this.drawer;
    },
  },
};
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped>
.link {
  cursor: pointer;
}
.link:hover {
  text-decoration: underline;
}
.read-more {
  color: blue;
}
</style>
