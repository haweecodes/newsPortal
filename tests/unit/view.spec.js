import Vuetify from 'vuetify';
import {
  shallowMount,
  createLocalVue,
} from '@vue/test-utils';
import Vuex from 'vuex';
import Home from '../../src/views/Home.vue';
import Detail from '../../src/views/Detail.vue';

import {
  state,
  mutations,
  actions,
} from '../../src/store/index';

import {
  formatDate,
} from '../../src/utils/helper';

const vuetify = new Vuetify();
const localVue = createLocalVue();

describe('test Home', () => {
  let store;

  beforeEach(() => {
    localVue.use(Vuex);
    store = new Vuex.Store({
      actions,
      state,
      mutations,
    });
  });
  it('test Home mount', () => {
    const wrapper = shallowMount(Home, {
      localVue,
      store,
      mocks: {
        $vuetify: {
          breakpoint: {},
        },
      },
      vuetify,
    });

    expect(wrapper.html()).toMatchSnapshot();
  });
});

describe('test Detail', () => {
  let store;

  beforeEach(() => {
    localVue.use(Vuex);
    localVue.filter('transformDate', formatDate);

    store = new Vuex.Store({
      actions,
      state,
    });
  });
  it('test Detail mount', () => {
    const wrapper = shallowMount(Detail, {
      localVue,
      store,
      mocks: {
        $vuetify: {
          breakpoint: {},
        },
      },
      vuetify,
    });

    expect(wrapper.html()).toMatchSnapshot();
  });
});
