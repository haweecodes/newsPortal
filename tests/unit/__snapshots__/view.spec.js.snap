// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Detail test Detail mount 1`] = `
<article>
  <v-btn-stub tag="button" activeclass="" icon="true" type="button">
    <v-icon-stub>mdi-arrow-left</v-icon-stub>
  </v-btn-stub>
  <h1 class="text-center ma-5"></h1>
  <p class="text-center ma-5"><strong>Written By:</strong></p>
  <p class="text-center ma-5"><strong>Published At: 2022-04-28</strong></p>
  <v-img-stub aspectratio="3" contain="true" options="[object Object]" position="center center" src="" transition="fade-transition"></v-img-stub>
  <p class="ma-5"></p>
</article>
`;

exports[`test Home test Home mount 1`] = `
<section class="h-100">
  <v-container-stub tag="div" fluid="true" class="h-100">
    <searchandfilter-stub></searchandfilter-stub>
    <!---->
    <loaderwrapper-stub loading="true"></loaderwrapper-stub>
    <!---->
    <v-speed-dial-stub bottom="true" fixed="true" right="true" transition="slide-y-reverse-transition" direction="top" openonhover="true">
      <v-btn-stub color="indigo" dark="true" tag="button" activeclass="" small="true" fab="true" type="button">
        <v-icon-stub>mdi-history</v-icon-stub>
      </v-btn-stub>
      <v-btn-stub color="red" dark="true" tag="button" activeclass="" small="true" fab="true" type="button"> 404 </v-btn-stub>
    </v-speed-dial-stub>
    <v-snackbar-stub color="red accent-2" tile="true" tag="div" value="" absolute="true" right="true" top="true" centered="true" contentclass="" timeout="-1" transition="v-snack-transition">


    </v-snackbar-stub>
  </v-container-stub>
</section>
`;
